### ISSUE TITLE
exp run: Is applying every time without dep changes

### ISSUE THREAD

**Issue by DavidGOrtega** [iterative/dvc#6084]
Hi!

Im running into an issue that probably is my mistake so I apologise in advance 🙏 
To follow up you just need to fork https://github.com/iterative/dvc-checkpoints-mnist

## Description

Every time I run ```dvc exp run``` I have 10 epochs even I have not changed any dependancy.

### Reproduce

```sh
git clone https://github.com/iterative/dvc-checkpoints-mnist
cd dvc-checkpoints-mnist
dvc init -f
dvc remote add -d myremote...
dvc exp run

git add dvc.yaml dvc.lock dvclive.json train.py
git commit -m '10 epochs'

dvc push
git push
```

Now if I repeat ```exp run``` without any change It should not run like repro, right? However in my case every single time it runs another 10 epochs.

<details><summary>LOG</summary>
<p>

```sh
Running stage 'train':                                                                                                                                 
> python3 train.py
Updating lock file 'dvc.lock'                                                                                                                          
Checkpoint experiment iteration 'd64afd6'.

file:///iterative/dvc-checkpoints-mnist/dvclive.html
Updating lock file 'dvc.lock'                                                                                                                          
Checkpoint experiment iteration '8f532c0'.

file:///iterative/dvc-checkpoints-mnist/dvclive.html
Updating lock file 'dvc.lock'                                                                                                                          
Checkpoint experiment iteration '9f5b61b'.

file:///iterative/dvc-checkpoints-mnist/dvclive.html
Updating lock file 'dvc.lock'                                                                                                                          
Checkpoint experiment iteration '1ba2514'.

file:///iterative/dvc-checkpoints-mnist/dvclive.html
Updating lock file 'dvc.lock'                                                                                                                          
Checkpoint experiment iteration 'd70a50f'.

file:///iterative/dvc-checkpoints-mnist/dvclive.html
Updating lock file 'dvc.lock'                                                                                                                          
Checkpoint experiment iteration '450cab5'.

file:///iterative/dvc-checkpoints-mnist/dvclive.html
Updating lock file 'dvc.lock'                                                                                                                          
Checkpoint experiment iteration 'a599f5e'.

file:///iterative/dvc-checkpoints-mnist/dvclive.html
Updating lock file 'dvc.lock'                                                                                                                          
Checkpoint experiment iteration '0043e6c'.

file:///iterative/dvc-checkpoints-mnist/dvclive.html
Updating lock file 'dvc.lock'                                                                                                                          
Checkpoint experiment iteration '18fda60'.

file:///iterative/dvc-checkpoints-mnist/dvclive.html
Updating lock file 'dvc.lock'                                                                                                                          
Checkpoint experiment iteration '01017c1'.

file:///iterative/dvc-checkpoints-mnist/dvclive.html

file:///iterative/dvc-checkpoints-mnist/dvclive.html
                                                                                                                                                       
To track the changes with git, run:

        git add train.py dvclive.json dvc.lock dvc.yaml
                                                                      
Reproduced experiment(s): exp-b434e
Experiment results have been applied to your workspace.

To promote an experiment to a Git branch run:

        dvc exp branch <exp>
```
</p>
</details>

### Expected

If I don't change any dependency exp run like repro should not run?

### Environment information

```console
DVC version: 2.1.0 (deb)
---------------------------------
Platform: Python 3.8.9 on Linux-4.19.76-linuxkit-x86_64-with-glibc2.7
Supports: All remotes
Cache types: hardlink, symlink
Cache directory: fuse.grpcfuse on grpcfuse
Caches: local
Remotes: s3
Workspace directory: fuse.grpcfuse on grpcfuse
Repo: dvc, git
```

### Extra 

As I curiosity if I do not git commit the list grows everytime.

<details><summary>dvc exp show</summary>
<p>

```
━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━┳━━━━━━━━┓
┃ Experiment              ┃ Created  ┃ step ┃ acc    ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━╇━━━━━━━━┩
│ workspace               │ -        │ 29   │ 0.9823 │
│ live                    │ 02:18 PM │ 19   │ 0.9831 │
│ │ ╓ 8ce5a03 [exp-fa8ae] │ 02:25 PM │ 29   │ 0.9823 │
│ │ ╟ 75cde7c             │ 02:25 PM │ 28   │ 0.9849 │
│ │ ╟ 8085755             │ 02:24 PM │ 27   │ 0.9857 │
│ │ ╟ 6e4c1c7             │ 02:24 PM │ 26   │ 0.9819 │
│ │ ╟ 23238d4             │ 02:24 PM │ 25   │ 0.9833 │
│ │ ╟ 9ec8896             │ 02:24 PM │ 24   │ 0.9831 │
│ │ ╟ 8ad79a9             │ 02:24 PM │ 23   │ 0.985  │
│ │ ╟ 137d8fd             │ 02:24 PM │ 22   │ 0.9844 │
│ │ ╟ 026104d             │ 02:23 PM │ 21   │ 0.9852 │
│ │ ╟ 3e4dca0             │ 02:23 PM │ 20   │ 0.984  │
│ │ ╟ 2ffa35b             │ 02:23 PM │ 19   │ 0.9827 │
│ │ ╟ d66a0a6             │ 02:22 PM │ 18   │ 0.9849 │
│ │ ╟ 916d1aa             │ 02:22 PM │ 17   │ 0.9857 │
│ │ ╟ d72cf34             │ 02:22 PM │ 16   │ 0.9845 │
│ │ ╟ f5dc213             │ 02:22 PM │ 15   │ 0.9831 │
│ │ ╟ 3d05cc0             │ 02:22 PM │ 14   │ 0.9831 │
│ │ ╟ 60d2f69             │ 02:22 PM │ 13   │ 0.9841 │
│ │ ╟ dab6a6f             │ 02:21 PM │ 12   │ 0.9846 │
│ │ ╟ 2413044             │ 02:21 PM │ 11   │ 0.9862 │
│ │ ╟ 822b2ee             │ 02:21 PM │ 10   │ 0.9839 │
│ │ ╟ 2361ca5             │ 02:20 PM │ 9    │ 0.9834 │
│ │ ╟ 45cb159             │ 02:20 PM │ 8    │ 0.9859 │
│ │ ╟ a68a635             │ 02:20 PM │ 7    │ 0.9864 │
│ │ ╟ 21a2fbd             │ 02:20 PM │ 6    │ 0.9834 │
│ │ ╟ ad05469             │ 02:19 PM │ 5    │ 0.9845 │
│ │ ╟ e0a0577             │ 02:19 PM │ 4    │ 0.985  │
│ │ ╟ 0ea9dc6             │ 02:19 PM │ 3    │ 0.9845 │
│ │ ╟ 35411f9             │ 02:19 PM │ 2    │ 0.9862 │
│ │ ╟ 8c6599b             │ 02:19 PM │ 1    │ 0.9848 │
│ ├─╨ 968da1a             │ 02:19 PM │ 0    │ 0.9855 │
└─────────────────────────┴──────────┴──────┴────────
```

</p>
</details>

**Comment by pmrowla** [comment]
This is expected behavior, `dvc-checkpoints-mnist` is a checkpoint example project. With checkpoints, every time you do `exp run`, the experiment is resumed "continue" from the last checkpoint. So in this case, it means each `exp run` will add 10 more epochs.

`exp run` will behave like `repro` (and not extend the experiment branch) for unchanged "regular" experiments that do not have checkpoints.

**Comment by shcheklein** [comment]
@jorgeorpinel I think we can improve our docs? Be explicit at the very beginning `With checkpoints, every time you do exp run, the experiment is resumed from the last checkpoint`.  (docs for `exp run` are long and a a bit too low level).

Help for the command is misleading and low level also:

`Reproduce complete or partial experiment pipelines.` 

It should be probably as simple as `Run or resume an experiment`

**Comment by DavidGOrtega** [comment]
@pmrowla so using checkpoints is not the choice to save checkpoints to resume training later on right? What would be the best choice for such a need?

**Comment by pmrowla** [comment]
@DavidGOrtega I don't really understand the question here, might be better for @dberenbaum to handle it?

The default behavior for `exp run` and checkpoints is to always resume training from the last generated checkpoint.

**Comment by dberenbaum** [comment]
I'm not sure I fully understand either. Is this what you are trying to do:

1. run an experiment for a few epochs
2. run a separate experiment with new params for a few epochs, starting from scratch
3. return to the first experiment and add some more epochs

**Comment by DavidGOrtega** [comment]
@dberenbaum not really I just want to setup checkpoints to resume training if by any reason it fails. Which is a common use case for checkpoints.

**Comment by dberenbaum** [comment]
Okay, I think I understand 🤞 .

Even though it might look like the repo is starting from scratch, every time you invoke `dvc exp run` it will add another 10 epochs to the existing checkpoints. You can check this by running `dvc exp run` multiple times and then running `dvc exp show` (you should see more than 10 checkpoints and continually increasing accuracy).

Quoting from https://discuss.dvc.org/t/experiments-with-checkpoint/744/4:

> > Also, in GitHub - iterative/dvc-checkpoints-mnist: Example of checkpoints in a DVC project training a simple convolutional neural net to classify MNIST data, the training consists of 10 epoch, but if you stop the training at epoch 5, and restart wouldn’t the model train for 15 epochs ?
> 
> Yes, that’s what will happen. In that example, 10 epochs is intended to be a starting point rather than a set number of epochs. In many cases, a checkpoints script might be an infinite loop, and the user will determine when to stop training based on the trajectory of the metrics. The checkpoints feature is intended for circumstances where you don’t know how long you want to train and want to keep adding iterations (and possibly go back if performance starts getting worse).

**Comment by dberenbaum** [comment]
By the way, @DavidGOrtega please follow up with what was confusing, because we are working on documenting checkpoints better right now.  cc @jorgeorpinel @iesahin

**Comment by DavidGOrtega** [comment]
@dberenbaum yes, I realised that we are adding checkpoints, the question is how I would just train X checkpoints like if I were doing it with TF, Caffe... Use Checkpoints is good mostly for:
 - resume training
 - go back to checkpoint to finetune or just when you realise that you are overfitting etc...

In CML would be fundamental to be able to continue training if the runner failed or got disposed before the training ended. 
However this "forever" nature of our checkpoint would end up in an endless training since there is no stoppage. Unless you check at least ```dvc exp show``` in the train.py or something

**Comment by DavidGOrtega** [comment]
Citing:
> Yes, that’s the idea.
In my institution, we have a supercomputing grid on which we can run special “besteffort” jobs, which can be stopped at any time but can be run on more computing nodes.

Indeed is a very common scenario to fail and train from the last checkpoint covering your expected epochs and not more

**Comment by dberenbaum** [comment]
@DavidGOrtega Guess I misunderstood again 😄 . Please take a look at the discussion in https://discuss.dvc.org/t/experiments-with-checkpoint/744/4 as it seems very similar to what you are saying.

The forever nature of it is intentional AFAIK (I wasn't here for its initial development) so that users can keep going until they are satisfied with the metrics (often it will take an unknown number of iterations). I agree that it's clunky as a retry mechanism like you seem to be wanting.

**Comment by DavidGOrtega** [comment]
> @DavidGOrtega Guess I misunderstood again 😄 . Please take a look at the discussion in https://discuss.dvc.org/t/experiments-with-checkpoint/744/4 as it seems very similar to what you are saying.

Im citing it so seems that I already read it 😉 

This is a must have scenario in CML to be able to use spot instances in example.

**Comment by dberenbaum** [comment]
Yup, will the solution in that thread work for you in the short term?

**Comment by DavidGOrtega** [comment]
@dberenbaum Im not sure whats the solution. Actually for sure it does not apply to me since I have just only one stage?

**Comment by dberenbaum** [comment]
You can track the epoch number as an output and have the code exit whenever it reaches that epoch.

I also opened up a discussion: https://github.com/iterative/dvc/discussions/6104.

**Comment by dberenbaum** [comment]
By the way, if you are willing to use `dvclive`, this will take care of the epoch tracking for you, and you just need to read in the `dvclive.json` file and have your script stop when the `step` value in that file reaches your desired endpoint.

> In that case, instead of iterating over a constant range of steps/epochs like in dvc-checkpoints-mnist, you want to track the step number by reading in dvclive.json (or whatever path you are using for dvclive) and iterate until that step number is reached, right?

**Comment by DavidGOrtega** [comment]
> dvclive.json file and have your script stop when the step value in that file reaches your desired endpoint.

thats a good idea. Only would work for python unfortunately, but I would be able to use another file tracked by DVC.

**Comment by jorgeorpinel** [comment]
> @jorgeorpinel I think we can improve our docs? 

Agreed ⌛ 

But note that this one example repo isn't used much in docs other than linked for a couple places: in https://dvc.org/doc/user-guide/experiment-management#checkpoints-in-source-code and bottom of https://dvc.org/doc/dvclive/dvclive-with-dvc . And it's README is explicit about this behavior (see https://github.com/iterative/dvc-checkpoints-mnist#experimenting). @DavidGOrtega may I ask how did you find about the repo? Just curious

But yes, regardless we can improve explanations about checkpoints to correct this expectation:

> If I don't change any dependency exp run like repro should not run?
